torch_xla_py/data_parallel.py

```diff
@@ -0,0 +1,166 @@
+from __future__ import division
+from __future__ import print_function
+
+from six import iteritems, itervalues
+import threading
+import torch
+import torch_xla
+import torch_xla_py.utils as xu
+import torch_xla_py.xla_model as xm
+import torch_xla_py.utils as xu
+import torch_xla_py.keyd_queue as kq
+
+
+class ThreadResult(object):
+
+  def __init__(self):
+    self.result = None
+
+
+class PerDeviceQueue(object):
+
+  def __init__(self, device, maxsize):
+    self.device = device
+    self.batch_number = 0
+    self.queue = kq.Queue(maxsize=maxsize)
+
+
+class PerDeviceLoader(object):
+
+  def __init__(self, loader, device):
+    self._loader = loader
+    self._device = device
+
+  def __iter__(self):
+    return self
+
+  def __next__(self):
+    return self.next()
+
+  def next(self):
+    item = self._loader.next(self._device)
+    if item is None:
+      raise StopIteration
+    return item
+
+
+class ParallelLoader(object):
+
+  def __init__(self,
+               loader,
+               devices,
+               batchdim=0,
+               loader_prefetch_size=8,
+               device_prefetch_size=4):
+    self._loader = loader
+    self._batch_size = None
+    self._devices = list(devices)
+    self._batchdim = batchdim
+    self._done = False
+    self._lock = threading.Lock()
+    self._loader_queue = kq.Queue(maxsize=loader_prefetch_size)
+    self._queues = dict()
+    for device in self._devices:
+      self._queues[device] = PerDeviceQueue(device, device_prefetch_size)
+    thread = threading.Thread(target=self._loader_worker)
+    thread.daemon = True
+    thread.start()
+    for dqueue in itervalues(self._queues):
+      thread = threading.Thread(target=self._worker, args=(dqueue,))
+      thread.daemon = True
+      thread.start()
+
+  def per_device_loader(self, device):
+    return PerDeviceLoader(self, device)
+
+  def next(self, device):
+    dqueue = self._queues[device]
+    return dqueue.queue.get()
+
+  def close(self):
+    self._done = True
+    for dqueue in itervalues(self._queues):
+      dqueue.queue.close()
+    self._loader_queue.close()
+
+  def _expand_sample_batch(self, data, target):
+    # TODO: Expand last sample,target to batch size
```

Comment:
I'm not sure how we'd do this, let the user specify the padding value? Could be something else than zero, at least in theory. We can find the correct one through graph analysis, longer term.

Reply:
Not critical, but many ways to do it.
1) Slice the last sample, broadcast in batchdim by (batch_size - last_batch_size), concat.
2) Slice the first (batch_size - last_batch_size), concat.
3) Keep batch-1, slice top (batch_size - last_batch_size), concat.